crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_map.rs

```diff
@@ -67,140 +64,91 @@ impl Violation for UnnecessaryMap {
                 return;
             }
 
-            let Some(argument) = helpers::first_argument_with_matching_function("map", func, args)
-            else {
+            let Some(result) = map_lambda_and_iterables(call, semantic) else {
                 return;
             };
 
-            let Expr::Lambda(ast::ExprLambda {
-                parameters, body, ..
-            }) = argument
-            else {
-                return;
-            };
-
-            if !lambda_has_expected_arity(parameters.as_deref(), body) {
-                return;
-            }
+            result
         }
-        ObjectType::Dict => {
-            // Only flag, e.g., `dict(map(lambda v: (v, v ** 2), values))`.
-            let [Expr::Call(ast::ExprCall {
-                func,
-                arguments: Arguments { args, keywords, .. },
-                ..
-            })] = args
-            else {
-                return;
-            };
 
-            if args.len() != 2 {
+        ObjectType::List | ObjectType::Set | ObjectType::Dict => {
+            if !arguments.keywords.is_empty() {
                 return;
             }
 
-            if !keywords.is_empty() {
-                return;
-            }
-
-            let Some(argument) = helpers::first_argument_with_matching_function("map", func, args)
-            else {
+            let [Expr::Call(inner_call)] = arguments.args.as_ref() else {
                 return;
             };
 
-            let Expr::Lambda(ast::ExprLambda {
-                parameters, body, ..
-            }) = argument
+            let Some((parameters, body, iterables)) =
+                map_lambda_and_iterables(inner_call, semantic)
             else {
                 return;
             };
 
-            let (Expr::Tuple(ast::ExprTuple { elts, .. }) | Expr::List(ast::ExprList { elts, .. })) =
-                body.as_ref()
-            else {
-                return;
-            };
+            if object_type == ObjectType::Dict {
+                let (Expr::Tuple(ast::ExprTuple { elts, .. })
+                | Expr::List(ast::ExprList { elts, .. })) = body
+                else {
+                    return;
+                };
 
-            if elts.len() != 2 {
-                return;
+                if elts.len() != 2 {
+                    return;
+                }
             }
 
-            if !lambda_has_expected_arity(parameters.as_deref(), body) {
-                return;
-            }
+            (parameters, body, iterables)
         }
     };
 
-    let mut diagnostic = Diagnostic::new(UnnecessaryMap { object_type }, expr.range());
+    // For example, (x+1 for x in (c:=a)) is invalid syntax
+    // so we can't suggest it.
+    if iterables
+        .iter()
+        .any(|iterable| any_over_expr(iterable, &|expr| expr.is_named_expr()))
+    {
+        return;
+    }
+
+    if iterables.iter().any(Expr::is_starred_expr) {
+        return;
+    }
+
```

Comment:
Nit: We could do this in a single iteration
```suggestion
    // For example, (x+1 for x in (c:=a)) is invalid syntax
    // so we can't suggest it.
    for iterable in &iterables {
        if iterable.is_starred_expr() {
            return;
        }

        if any_over_expr(iterable, &|expr| expr.is_named_expr()) {
            return
        }
    }
```

or 

```suggestion
    // For example, (x+1 for x in (c:=a)) is invalid syntax
    // so we can't suggest it.
    if iterables
        .iter()
        .any(|iterable| iterable.is_starred_expr() || any_over_expr(iterable, &|expr| expr.is_named_expr()))
    {
        return;
    }
```